### PR TITLE
Add support for version upgrades

### LINKED ISSUE #4141: Concept for the upgrade 5.3 -> 5.4
Once available, the upgrade should be as simple as possible, without requiring users to manually edit anything. Ideally we run some check for the requirements and make sure that all outstanding updates are installed before running the upgrade.

Potential behavior:
1.  Message on the package list informs of the new version, with a link or button to upgrade.
2. Clicking the link should run a check for the system requirements.
3. If the check was successful, check for outstanding updates.
4. If no outstanding updates were found, set a session variable to override the package servers for 5.4.
5. Run a "check for updates", which will redirect the user.

Outstanding questions:
* Do we need to manually handle connection issues preventing communications with the package server?
* Do we require adjustments for the Cloud instances?

/cc @TimWolla

### PR DESCRIPTION
**Pull request by TimWolla**
Resolves #4141

### CODE REVIEW

**Review by dtdesign — APPROVED** [review]
LGTM.

Can we get a check for PHP/MySQl requirements built into the activation process for the upgrade?

**Comment by TimWolla** [comment]
> Can we get a check for PHP/MySQl requirements built into the activation process for the upgrade?

Sure. That's why this still is a draft with a general PoC :smiley:

File: wcfsetup/install/lang/de.xml [hunk 1]
```diff
@@ -1990,6 +1990,10 @@ Die Datenbestände werden sorgfältig gepflegt, aber es ist nicht ausgeschlossen
 		Unbekannter Fehler.
 	{/if}
 {/implode}]]></item>
+		<item name="wcf.acp.package.enableUpgradeOverride"><![CDATA[WoltLab Suite-Upgrade]]></item>
+		<item name="wcf.acp.package.enableUpgradeOverride.enable"><![CDATA[Upgrade freigeben]]></item>
+		<item name="wcf.acp.package.availableUpgradeVersion"><![CDATA[Ein WoltLab Suite-Upgrade zur Verfügung. <a href="{link controller='PackageEnableUpgradeOverride' application='wcf'}{/link}">System prüfen und Upgrade freigeben</a>.]]></item>
```
Comment (dtdesign) [inline wcfsetup/install/lang/de.xml:1995]:
```suggestion
		<item name="wcf.acp.package.availableUpgradeVersion"><![CDATA[Ein WoltLab Suite-Upgrade steht zur Verfügung. <a href="{link controller='PackageEnableUpgradeOverride' application='wcf'}{/link}">System prüfen und Upgrade freigeben</a>.]]></item>
```

File: wcfsetup/install/lang/de.xml [hunk 2]
```diff
@@ -1990,6 +1990,10 @@ Die Datenbestände werden sorgfältig gepflegt, aber es ist nicht ausgeschlossen
...
+		<item name="wcf.acp.package.upgradeOverrideEnabled"><![CDATA[Die Installation eines WoltLab Suite-Upgrades wurde freigegeben. Die Durchführung von Updates wird {if LANGUAGE_USE_INFORMAL_VARIANT}deine{else}Ihre{/if} Community auf eine neue Version von WoltLab Suite upgraden. <a href="{link controller='PackageEnableUpgradeOverride' application='wcf'}{/link}">Freigabe zurückziehen</a>.]]></item>
```
Comment (dtdesign) [inline wcfsetup/install/lang/de.xml:1996]:
```suggestion
		<item name="wcf.acp.package.upgradeOverrideEnabled"><![CDATA[Die Installation eines WoltLab Suite-Upgrades wurde freigegeben. Die Durchführung von Updates wird {if LANGUAGE_USE_INFORMAL_VARIANT}deine{else}Ihre{/if} Community auf eine neue Version von WoltLab Suite aktualisieren. <a href="{link controller='PackageEnableUpgradeOverride' application='wcf'}{/link}">Freigabe zurückziehen</a>.]]></item>
```

**Review by dtdesign — CHANGES_REQUESTED** [review]
LGTM

File: wcfsetup/install/files/lib/system/package/PackageUpdateDispatcher.class.php [hunk 3]
```diff
@@ -153,6 +153,16 @@ protected function getPackageUpdateXML(PackageUpdateServer $updateServer, $force
 		
 		$request = new HTTPRequest($updateServer->getListURL($forceHTTP), $settings);
 		
+		$requestedVersion = \wcf\getMinorVersion();
+		if (PackageUpdateServer::isUpgradeOverrideEnabled()) {
+			$requestedVersion = WCF::AVAILABLE_UPGRADE_VERSION;
+		}
+		
+		$request->addHeader(
+			'requested-woltlab-suite-version',
```
Comment (jens1o) [inline wcfsetup/install/files/lib/system/package/PackageUpdateDispatcher.class.php:162]:
```suggestion
			'x-requested-woltlab-suite-version',
```
?

Comment (TimWolla) [reply wcfsetup/install/files/lib/system/package/PackageUpdateDispatcher.class.php:162]: No, RFC 6648.

File: wcfsetup/install/files/lib/acp/page/PackageListPage.class.php [hunk 4]
```diff
@@ -71,10 +72,18 @@ public function readParameters() {
...
+		// During the RC phase the upgrade should only be offered when the maintenance mode is enabled.
+		if (!\OFFLINE) {
+			$availableUpgradeVersion = null;
+		}
+
 		WCF::getTPL()->assign([
 			'recentlyDisabledCustomValues' => LanguageFactory::getInstance()->countRecentlyDisabledCustomValues(),
-			'packageID' => $this->packageID
+			'packageID' => $this->packageID,
+			'availableUpgradeVersion' => WCF::AVAILABLE_UPGRADE_VERSION,
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/acp/page/PackageListPage.class.php:85]:
```suggestion
			'availableUpgradeVersion' => $availableUpgradeVersion,
```

File: wcfsetup/install/files/acp/templates/packageUpdateServerList.tpl [hunk 5]
```diff
@@ -21,6 +21,16 @@
 	</nav>
 </header>
 
+{if $__wcf->session->getPermission('admin.configuration.package.canUpdatePackage')}
+	{if $availableUpgradeVersion !== null}
+		{if $upgradeOverrideEnabled}
+			<p class="info">{lang}wcf.acp.package.upgradeOverrideEnabled{/lang}</p>
+		{else}
+			<p class="info">{lang}wcf.acp.package.availableUpgradeVersion{/lang}</p>
```
Comment (TimWolla) [inline wcfsetup/install/files/acp/templates/packageUpdateServerList.tpl:29]:
```suggestion
			<p class="info">{lang version=$availableUpgradeVersion}wcf.acp.package.availableUpgradeVersion{/lang}</p>
```

File: wcfsetup/install/files/acp/templates/packageUpdateServerList.tpl [hunk 6]
```diff
@@ -21,6 +21,16 @@
 	</nav>
 </header>
 
+{if $__wcf->session->getPermission('admin.configuration.package.canUpdatePackage')}
+	{if $availableUpgradeVersion !== null}
+		{if $upgradeOverrideEnabled}
+			<p class="info">{lang}wcf.acp.package.upgradeOverrideEnabled{/lang}</p>
```
Comment (TimWolla) [inline wcfsetup/install/files/acp/templates/packageUpdateServerList.tpl:27]:
```suggestion
			<p class="info">{lang version=$availableUpgradeVersion}wcf.acp.package.upgradeOverrideEnabled{/lang}</p>
```

File: wcfsetup/install/files/lib/acp/page/PackageUpdateServerListPage.class.php [hunk 7]
```diff
@@ -47,4 +49,16 @@ public function readObjects() {
...
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function assignVariables() {
+		parent::assignVariables();
+		
+		WCF::getTPL()->assign([
+			'availableUpgradeVersion' => WCF::AVAILABLE_UPGRADE_VERSION,
+			'upgradeOverrideEnabled' => PackageUpdateServer::isUpgradeOverrideEnabled(),
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/acp/page/PackageUpdateServerListPage.class.php:61]: Should this look like `PackageListPage`?

**Pull request merged**